lxd/cluster/membership.go

```diff
@@ -401,12 +414,48 @@ func Join(state *state.State, gateway *Gateway, networkCert *shared.CertInfo, se
 	// Re-initialize the gateway. This will create a new raft factory an
 	// dqlite driver instance, which will be exposed over gRPC by the
 	// gateway handlers.
+	oldCert := gateway.networkCert
 	gateway.networkCert = networkCert
 	err = gateway.init(false)
 	if err != nil {
 		return fmt.Errorf("Failed to re-initialize gRPC SQL gateway: %w", err)
 	}
 
+	reverter.Add(func() {
+		err = state.DB.Node.Transaction(context.TODO(), func(ctx context.Context, tx *db.NodeTx) error {
+			return tx.ReplaceRaftNodes([]db.RaftNode{})
+		})
+		if err != nil {
+			logger.Error("Failed to clear local raft node records after cluster join error", logger.Ctx{"error": err})
+			return
+		}
+
+		err = gateway.Shutdown()
+		if err != nil {
+			logger.Error("Failed to shutdown gateway after cluster join error", logger.Ctx{"error": err})
+			return
+		}
+
+		err = os.RemoveAll(state.OS.GlobalDatabaseDir())
```

Comment:
What happens if the joining server already had data in it?

Reply:
We`re clearing out the database files so any data would be gone. I thought about preserving the old database files, but then considering that we prompt the user with a warning during init that joining a cluster will wipe existing data, I went went with the simpler choice since the user already consented to sacrificing the data. 

We could just move the old files to a tmpdir and defer the removal.

Reply:
For context (since it's hidden in the diff), prior to this change we were already wiping all data on the existing node:

https://github.com/canonical/lxd/blob/0aabb554848e5bc708260a9390eea529551ada0b/lxd/cluster/membership.go#L396-L399